crates/driver/src/domain/competition/mod.rs

```diff
@@ -84,47 +86,102 @@ impl Competition {
             }
         });
 
-        // Verify the solutions, encoding them into settlements.
+        // Encode the solutions into settlements.
         let settlements = join_all(solutions.map(|solution| async move {
-            tracing::trace!(id = ?solution.id, "verifying");
+            tracing::trace!(id = ?solution.id, "encoding settlement");
             (
                 solution.id,
-                solution.verify(&self.eth, &self.simulator, auction).await,
+                solution.encode(auction, &self.eth, &self.simulator).await,
             )
         }))
         .await;
 
-        // Filter out solutions that failed verification.
-        let settlements = settlements.into_iter().filter_map(|(id, result)| {
-            result
-                .tap_err(|err| {
-                    tracing::info!(?err, ?id, "discarding solution: failed verification")
-                })
-                .ok()
-        });
-
-        // Score the solutions.
-        let scores = join_all(settlements.map(|settlement| async move {
-            tracing::trace!(id = ?settlement.solution.id, "scoring");
-            (settlement.score(&self.eth, auction).await, settlement)
-        }))
-        .await;
-
-        // Filter out solutions which failed scoring.
-        let scores = scores
+        // Filter out solutions that failed to encode.
+        let mut settlements = settlements
             .into_iter()
-            .filter_map(|(result, settlement)| {
+            .filter_map(|(id, result)| {
                 result
                     .tap_err(|err| {
-                        tracing::info!(?err, id = ?settlement.solution.id, "discarding solution: failed scoring")
+                        tracing::info!(?err, ?id, "discarding solution: settlement encoding failed")
                     })
                     .ok()
-                    .map(|score| (score, settlement))
-            });
+            })
+            .collect_vec();
+
+        // TODO(#1483): parallelize this
+        // TODO(#1480): more optimal approach for settlement merging
+
+        // Merge the settlements in random order.
+        settlements.shuffle(&mut rand::thread_rng());
+        // The merged settlements in their final form.
+        let mut results = Vec::new();
+        while let Some(settlement) = settlements.pop() {
+            // Has [`settlement`] been merged into another settlement?
+            let mut merged = false;
+            // Try to merge [`settlement`] into some other settlement.
+            for other in settlements.iter_mut() {
+                match other.merge(&settlement, &self.eth, &self.simulator).await {
+                    Ok(m) => {
+                        tracing::debug!(
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "merged solutions"
+                        );
+                        *other = m;
```

Comment:
This is neat, so `settlements` vector keeps the "most merged" settlements until they can't be merged anymore and pushed into `results` vector.

One worry I have is that driver will actually miss solving deadlines because of this. Can we add a timeout to the "merge loop" so that we only keep on trying to merge for as long as we have time, and then just bail out and pick the best we've got once we've hit the deadline? (Can be done in a follow up).

Reply:
https://github.com/cowprotocol/services/issues/1498. Also, I added your summary as a comment here: https://github.com/cowprotocol/services/pull/1490/commits/f0393f9a60770bd7758ea51e743ca7d080fb8ecd